**Instructions**

1. Defining business objective
2. Extracting the data
3. Data cleaning, wrangling & EDA
    - On the **categorical** columns in the dataset:
        - Checking for null values in all the columns.
        - Excluding the following variables by looking at the definitions. Create a new empty list called drop_list. We will append this list and then drop all the columns in this list later: `OSOURCE`, `ZIP`.
        - Identifying columns that have over 85% missing values and removing them.
        - Reducing the number of categories in the column `GENDER`. The column should only have either "M" for males, "F" for females, and "other" for all the rest. 
    - On the **numerical** columns:
        - Checking for null values in the numerical columns.
        - Cleaning the columns GEOCODE2, WEALTH1, ADI, DMA,and MSA.
4. Checking accuracy
5. Building a model
6. Improving the model
7. Results

**Additional Information on Dataset**

- Large number of features: The data set has over 450 features. Hence selecting the right features for the model is very critical and at the same time it is not easy as the same traditional ways of removing features is not effective given the large number of features. Apart from feature selection, feature extraction (creating your own features using the existing features) is also not easy in this case.

- Sparsity: There are a lot of features with a large number of null values.

- Data imbalance: For developing a classification, there is a huge imbalance in the training dataset with only approximately 5000 values for one category as compared to over 95,000 instances for the other category.

In [1]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)
import warnings
warnings.filterwarnings('ignore')
import matplotlib.pyplot as plt
import seaborn as sns 

## 1. Defining business objective

To make their future direct marketing efforts more cost-effective, the organization, "Healthcare for All," wants to create a model that will help them maximize the overall revenue from future mailings. They want to develop a model that predicts the amount of money each donor is likely to give, rather than just classifying whether they will respond or not.
The organization is interested in a specific group called "lapsed" donors. These are people who made a donation between 13 to 24 months ago but haven't donated since then.
The organization observed a reverse relationship between the likelihood of a donor responding and the amount of money they give. This means that if they use a simple model to predict who will respond, they will mostly attract donors who give very small amounts.

The mailing was sent out in June 2018. All information included in the file (excluding the giving history date fields) is reflective of behavior before June 1997.

## 2. Extracting the data

In [2]:
df = pd.read_csv('learningSet.csv')

In [3]:
df.shape

(95412, 481)

It's a very large dataset, we have 481 features.

In [4]:
df.head()

ODATEDW OSOURCE  TCODE STATE    ZIP MAILCODE PVASTATE   DOB NOEXCH  \
0     8901     GRI      0    IL  61081                    3712      0   
1     9401     BOA      1    CA  91326                    5202      0   
2     9001     AMH      1    NC  27017                       0      0   
3     8701     BRY      0    CA  95953                    2801      0   
4     8601              0    FL  33176                    2001      0   

  RECINHSE RECP3 RECPGVG RECSWEEP MDMAUD DOMAIN CLUSTER   AGE AGEFLAG  \
0                                   XXXX     T2      36  60.0           
1                                   XXXX     S1      14  46.0       E   
2                                   XXXX     R2      43   NaN           
3                                   XXXX     R2      44  70.0       E   
4        X     X                    XXXX     S2      16  78.0       E   

  HOMEOWNR CHILD03 CHILD07 CHILD12 CHILD18  NUMCHLD  INCOME GENDER  WEALTH1  \
0                                               NaN     NaN      F      NaN   
1        H                               M      1.0     6.0      M      9.0   
2        U                                      NaN     3.0      M      1.0   
3        U                                      NaN     1.0      F      4.0   
4        H                                      1.0     3.0      F      2.0   

   HIT  MBCRAFT  MBGARDEN  MBBOOKS  MBCOLECT  MAGFAML  MAGFEM  MAGMALE  \
0    0      NaN       NaN      NaN       NaN      NaN     NaN      NaN   
1   16      0.0       0.0      3.0       1.0      1.0     1.0      0.0   
2    2      0.0       0.0      1.0       0.0      0.0     0.0      0.0   
3    2      0.0       0.0      0.0       0.0      0.0     0.0      0.0   
4   60      1.0       0.0      9.0       0.0      4.0     1.0      0.0   

   PUBGARDN  PUBCULIN  PUBHLTH  PUBDOITY  PUBNEWFN  PUBPHOTO  PUBOPP DATASRCE  \
0       NaN       NaN      NaN       NaN       NaN       NaN     NaN            
1       0.0       0.0      2.0       0.0       3.0       0.0     0.0        3   
2       0.0       0.0      1.0       0.0       0.0       0.0     0.0        3   
3       0.0       0.0      0.0       0.0       0.0       0.0     1.0        3   
4       0.0       0.0      4.0       0.0       1.0       0.0     1.0        3   

   MALEMILI  MALEVET  VIETVETS  WWIIVETS  LOCALGOV  STATEGOV  FEDGOV SOLP3  \
0         0       39        34        18        10         2       1         
1         0       15        55        11         6         2       1         
2         0       20        29        33         6         8       1         
3         0       23        14        31         3         0       3         
4         1       28         9        53        26         3       2         

  SOLIH MAJOR  WEALTH2 GEOCODE COLLECT1 VETERANS BIBLE CATLG HOMEE PETS  \
0                  5.0                                                    
1                  9.0       2                                            
2                  1.0                                                    
3                  0.0                                                    
4    12            NaN                               Y     Y              

  CDPLAY STEREO PCOWNERS PHOTO CRAFTS FISHER GARDENIN BOATS WALKER KIDSTUFF  \
0                                                                             
1                                                                             
2                                                                             
3                                                                             
4             Y                     Y               Y            Y            

  CARDS PLATES LIFESRC PEPSTRFL  POP901  POP902  POP903  POP90C1  POP90C2  \
0                             X     992     264     332        0       35   
1                                  3611     940     998       99        0   
2                             X    7001    2040    2669        0        2   
3              

`TARGET_B` and `TARGET_D` could both be our target. 
- `TARGET_B`: donors that responded to a mailing ('1') and those that didn't ('0')
- `TARGET_D`: total amount that has been given per donor

In [5]:
df['TARGET_B'].value_counts()

0    90569
1     4843
Name: TARGET_B, dtype: int64

There is a huge imbalance towards those donors that did not respond in the past.

In [6]:
df['TARGET_D'].describe()

count    95412.000000
mean         0.793073
std          4.429725
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max        200.000000
Name: TARGET_D, dtype: float64

In [7]:
df['TARGET_D'].value_counts()

0.00     90569
10.00      941
15.00      591
20.00      577
5.00       503
         ...  
18.25        1
10.70        1
2.50         1
16.87        1
44.21        1
Name: TARGET_D, Length: 71, dtype: int64

The total amount donated varies heavily since many adressees haven't donated anything at all.

In [8]:
df.corr()['TARGET_D'].sort_values(ascending=False)

TARGET_D    1.000000
TARGET_B    0.774232
RAMNT_4     0.268811
RAMNT_21    0.099339
RAMNT_9     0.090168
              ...   
RDATE_3    -0.125194
RDATE_5    -0.220455
RAMNT_5    -0.272147
ADATE_5          NaN
ADATE_15         NaN
Name: TARGET_D, Length: 407, dtype: float64

In [9]:
df.corr()['TARGET_B'].sort_values(ascending=False)

TARGET_B    1.000000
TARGET_D    0.774232
RFA_2F      0.072311
CARDGIFT    0.054027
NGIFTALL    0.050896
              ...   
RAMNT_3    -0.095351
RDATE_3    -0.126060
RAMNT_5    -0.380296
ADATE_5          NaN
ADATE_15         NaN
Name: TARGET_B, Length: 407, dtype: float64

There are no high correlations of features with the targets. This will make it difficult to predict `TARGET_D` in a linear regression. `TARGET_B`, on the other hand, we would predict with a classification.

## Data cleaning, wrangling & EDA

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95412 entries, 0 to 95411
Columns: 481 entries, ODATEDW to GEOCODE2
dtypes: float64(97), int64(310), object(74)
memory usage: 350.1+ MB


Checking for null values in all the columns.

In [11]:
df.isna().sum()

ODATEDW       0
OSOURCE       0
TCODE         0
STATE         0
ZIP           0
           ... 
MDMAUD_R      0
MDMAUD_F      0
MDMAUD_A      0
CLUSTER2    132
GEOCODE2    132
Length: 481, dtype: int64

In [12]:
empty_columns = df.columns[df.eq(' ').any()]
empty_columns

Index(['OSOURCE', 'MAILCODE', 'PVASTATE', 'NOEXCH', 'RECINHSE', 'RECP3',
       'RECPGVG', 'RECSWEEP', 'DOMAIN', 'CLUSTER', 'AGEFLAG', 'HOMEOWNR',
       'CHILD03', 'CHILD07', 'CHILD12', 'CHILD18', 'GENDER', 'DATASRCE',
       'SOLP3', 'SOLIH', 'MAJOR', 'GEOCODE', 'COLLECT1', 'VETERANS', 'BIBLE',
       'CATLG', 'HOMEE', 'PETS', 'CDPLAY', 'STEREO', 'PCOWNERS', 'PHOTO',
       'CRAFTS', 'FISHER', 'GARDENIN', 'BOATS', 'WALKER', 'KIDSTUFF', 'CARDS',
       'PLATES', 'LIFESRC', 'PEPSTRFL', 'RFA_3', 'RFA_4', 'RFA_5', 'RFA_6',
       'RFA_7', 'RFA_8', 'RFA_9', 'RFA_10', 'RFA_11', 'RFA_12', 'RFA_13',
       'RFA_14', 'RFA_15', 'RFA_16', 'RFA_17', 'RFA_18', 'RFA_19', 'RFA_20',
       'RFA_21', 'RFA_22', 'RFA_23', 'RFA_24', 'GEOCODE2'],
      dtype='object')

But there are a lot of columns with empty values. Let's replace ' ' with NaN.

In [13]:
# Removing 'MAILCODE' column because here ' ' stands for 'address ok'.
mailcode_column = df['MAILCODE']
df.drop('MAILCODE', axis=1, inplace=True)
df = df.replace(' ', np.nan)
df = pd.concat([df, mailcode_column], axis=1)
df.head()

ODATEDW OSOURCE  TCODE STATE    ZIP PVASTATE   DOB NOEXCH RECINHSE RECP3  \
0     8901     GRI      0    IL  61081      NaN  3712      0      NaN   NaN   
1     9401     BOA      1    CA  91326      NaN  5202      0      NaN   NaN   
2     9001     AMH      1    NC  27017      NaN     0      0      NaN   NaN   
3     8701     BRY      0    CA  95953      NaN  2801      0      NaN   NaN   
4     8601     NaN      0    FL  33176      NaN  2001      0        X     X   

  RECPGVG RECSWEEP MDMAUD DOMAIN CLUSTER   AGE AGEFLAG HOMEOWNR CHILD03  \
0     NaN      NaN   XXXX     T2      36  60.0     NaN      NaN     NaN   
1     NaN      NaN   XXXX     S1      14  46.0       E        H     NaN   
2     NaN      NaN   XXXX     R2      43   NaN     NaN        U     NaN   
3     NaN      NaN   XXXX     R2      44  70.0       E        U     NaN   
4     NaN      NaN   XXXX     S2      16  78.0       E        H     NaN   

  CHILD07 CHILD12 CHILD18  NUMCHLD  INCOME GENDER  WEALTH1  HIT  MBCRAFT  \
0     NaN     NaN     NaN      NaN     NaN      F      NaN    0      NaN   
1     NaN     NaN       M      1.0     6.0      M      9.0   16      0.0   
2     NaN     NaN     NaN      NaN     3.0      M      1.0    2      0.0   
3     NaN     NaN     NaN      NaN     1.0      F      4.0    2      0.0   
4     NaN     NaN     NaN      1.0     3.0      F      2.0   60      1.0   

   MBGARDEN  MBBOOKS  MBCOLECT  MAGFAML  MAGFEM  MAGMALE  PUBGARDN  PUBCULIN  \
0       NaN      NaN       NaN      NaN     NaN      NaN       NaN       NaN   
1       0.0      3.0       1.0      1.0     1.0      0.0       0.0       0.0   
2       0.0      1.0       0.0      0.0     0.0      0.0       0.0       0.0   
3       0.0      0.0       0.0      0.0     0.0      0.0       0.0       0.0   
4       0.0      9.0       0.0      4.0     1.0      0.0       0.0       0.0   

   PUBHLTH  PUBDOITY  PUBNEWFN  PUBPHOTO  PUBOPP DATASRCE  MALEMILI  MALEVET  \
0      NaN       NaN       NaN       NaN     NaN      NaN         0       39   
1      2.0       0.0       3.0       0.0     0.0        3         0       15   
2      1.0       0.0       0.0       0.0     0.0        3         0       20   
3      0.0       0.0       0.0       0.0     1.0        3         0       23   
4      4.0       0.0       1.0       0.0     1.0        3         1       28   

   VIETVETS  WWIIVETS  LOCALGOV  STATEGOV  FEDGOV SOLP3 SOLIH MAJOR  WEALTH2  \
0        34        18        10         2       1   NaN   NaN   NaN      5.0   
1        55        11         6         2       1   NaN   NaN   NaN      9.0   
2        29        33         6         8       1   NaN   NaN   NaN      1.0   
3        14        31         3         0       3   NaN   NaN   NaN      0.0   
4         9        53        26         3       2   NaN    12   NaN      NaN   

  GEOCODE COLLECT1 VETERANS BIBLE CATLG HOMEE PETS CDPLAY STEREO PCOWNERS  \
0     NaN      NaN      NaN   NaN   NaN   NaN  NaN    NaN    NaN      NaN   
1       2      NaN      NaN   NaN   NaN   NaN  NaN    NaN    NaN      NaN   
2     NaN      NaN      NaN   NaN   NaN   NaN  NaN    NaN    NaN      NaN   
3     NaN      NaN      NaN   NaN   NaN   NaN  NaN    NaN    NaN      NaN   
4     NaN      NaN      NaN     Y     Y   NaN  NaN    NaN      Y      NaN   

  PHOTO CRAFTS FISHER GARDENIN BOATS WALKER KIDSTUFF CARDS PLATES LIFESRC  \
0   NaN    NaN    NaN      NaN   NaN    NaN      NaN   NaN    NaN     NaN   
1   NaN    NaN    NaN      NaN   NaN    NaN      NaN   NaN    NaN     NaN   
2   NaN    NaN    NaN      NaN   NaN    NaN      NaN   NaN    NaN     NaN   
3   NaN    NaN    NaN      NaN   NaN    NaN      NaN   NaN    NaN     NaN   
4   NaN      Y    NaN        Y   NaN      Y      NaN     Y    NaN       3   

  PEPSTRFL  POP901  POP902  POP903  POP90C1  POP90C2  POP90C3  POP90C4  \
0        X     992     264     332        0       35       65       47   
1      NaN    3611     940     998       99        0        0       50   
2        X    7001    2040   

Before cleaning, let's separate in categoricals, numericals and Y for the targets.

In [14]:
Y = df[['TARGET_B', 'TARGET_D']]
Y.sample(5)

TARGET_B  TARGET_D
92684         0       0.0
61902         0       0.0
83115         0       0.0
19112         0       0.0
8873          0       0.0

In [15]:
numerical = df.select_dtypes(np.number)
numerical = numerical.drop(columns = ['TARGET_B', 'TARGET_D'])
numerical.head()

ODATEDW  TCODE   DOB   AGE  NUMCHLD  INCOME  WEALTH1  HIT  MBCRAFT  \
0     8901      0  3712  60.0      NaN     NaN      NaN    0      NaN   
1     9401      1  5202  46.0      1.0     6.0      9.0   16      0.0   
2     9001      1     0   NaN      NaN     3.0      1.0    2      0.0   
3     8701      0  2801  70.0      NaN     1.0      4.0    2      0.0   
4     8601      0  2001  78.0      1.0     3.0      2.0   60      1.0   

   MBGARDEN  MBBOOKS  MBCOLECT  MAGFAML  MAGFEM  MAGMALE  PUBGARDN  PUBCULIN  \
0       NaN      NaN       NaN      NaN     NaN      NaN       NaN       NaN   
1       0.0      3.0       1.0      1.0     1.0      0.0       0.0       0.0   
2       0.0      1.0       0.0      0.0     0.0      0.0       0.0       0.0   
3       0.0      0.0       0.0      0.0     0.0      0.0       0.0       0.0   
4       0.0      9.0       0.0      4.0     1.0      0.0       0.0       0.0   

   PUBHLTH  PUBDOITY  PUBNEWFN  PUBPHOTO  PUBOPP  MALEMILI  MALEVET  VIETVETS  \
0      NaN       NaN       NaN       NaN     NaN         0       39        34   
1      2.0       0.0       3.0       0.0     0.0         0       15        55   
2      1.0       0.0       0.0       0.0     0.0         0       20        29   
3      0.0       0.0       0.0       0.0     1.0         0       23        14   
4      4.0       0.0       1.0       0.0     1.0         1       28         9   

   WWIIVETS  LOCALGOV  STATEGOV  FEDGOV  WEALTH2  POP901  POP902  POP903  \
0        18        10         2       1      5.0     992     264     332   
1        11         6         2       1      9.0    3611     940     998   
2        33         6         8       1      1.0    7001    2040    2669   
3        31         3         0       3      0.0     640     160     219   
4        53        26         3       2      NaN    2520     627     761   

   POP90C1  POP90C2  POP90C3  POP90C4  POP90C5  ETH1  ETH2  ETH3  ETH4  ETH5  \
0        0       35       65       47       53    92     1     0     0    11   
1       99        0        0       50       50    67     0     0    31     6   
2        0        2       98       49       51    96     2     0     0     2   
3        0        8       92       54       46    61     0     0    11    32   
4       99        0        0       46       54     2    98     0     0     1   

   ETH6  ETH7  ETH8  ETH9  ETH10  ETH11  ETH12  ETH13  ETH14  ETH15  ETH16  \
0     0     0     0     0      0      0      0     11      0      0      0   
1     4     2     6     4     14      0      0      2      0      1      4   
2     0     0     0     0      0      0      0      2      0      0      0   
3     6     2     0     0      0      0      0     31      0      0      1   
4     0     0     0     0      0      0      0      0      0      0      0   

   AGE901  AGE902  AGE903  AGE904  AGE905  AGE906  AGE907  CHIL1  CHIL2  \
0      39      48      51      40      50      54      25     31     42   
1      34      41      43      32      42      45      32     33     46   
2      35      43      46      37      45      49      23     35     40   
3      32      40      44      34      43      47      25     45     35   
4      33      45      50      36      46      50      27     34     43   

   CHIL3  AGEC1  AGEC2  AGEC3  AGEC4  AGEC5  AGEC6  AGEC7  CHILC1  CHILC2  \
0     27     11     14     18     17     13     11     15      12      11   
1     21     13     14     33     23     10      4      2      11      16   
2     25     13     20     19     16     13     10      8      15      14   
3     20     15     25     17     17     12      7      7      20      17   
4     23     14     21     13     15     20     12      5      13      15   

   CHILC3  CHILC4  CHILC5  HHAGE1  HHAGE2  HHAGE3  HHN1  HHN2  HHN3  HHN4  \
0      34      25      18      26      10      23    18    33    49    28   
1      36      22      15      12       1       5     4    21    75    55   
2      30      22      19      25      10   

The numerical features provide some personal data, the donor's neighbourhood, economic & job situation, and some information on the donation history. 

In [16]:
categorical = df.select_dtypes('object')
categorical.columns

Index(['OSOURCE', 'STATE', 'ZIP', 'PVASTATE', 'NOEXCH', 'RECINHSE', 'RECP3',
       'RECPGVG', 'RECSWEEP', 'MDMAUD', 'DOMAIN', 'CLUSTER', 'AGEFLAG',
       'HOMEOWNR', 'CHILD03', 'CHILD07', 'CHILD12', 'CHILD18', 'GENDER',
       'DATASRCE', 'SOLP3', 'SOLIH', 'MAJOR', 'GEOCODE', 'COLLECT1',
       'VETERANS', 'BIBLE', 'CATLG', 'HOMEE', 'PETS', 'CDPLAY', 'STEREO',
       'PCOWNERS', 'PHOTO', 'CRAFTS', 'FISHER', 'GARDENIN', 'BOATS', 'WALKER',
       'KIDSTUFF', 'CARDS', 'PLATES', 'LIFESRC', 'PEPSTRFL', 'RFA_2', 'RFA_3',
       'RFA_4', 'RFA_5', 'RFA_6', 'RFA_7', 'RFA_8', 'RFA_9', 'RFA_10',
       'RFA_11', 'RFA_12', 'RFA_13', 'RFA_14', 'RFA_15', 'RFA_16', 'RFA_17',
       'RFA_18', 'RFA_19', 'RFA_20', 'RFA_21', 'RFA_22', 'RFA_23', 'RFA_24',
       'RFA_2R', 'RFA_2A', 'MDMAUD_R', 'MDMAUD_F', 'MDMAUD_A', 'GEOCODE2',
       'MAILCODE'],
      dtype='object')

The categorical features provide information on type & origin of record, the donor's living location, family status, other personal data and interests as well as on frequency and amount given (RFA) and major donors (MDMAUD). 

### Lab Machine Learning Revisited

#### On the **categorical** columns in the dataset:

**Checking for null values in all the columns.**

In [17]:
categorical.isna().sum()

OSOURCE       928
STATE           0
ZIP             0
PVASTATE    93954
NOEXCH          7
            ...  
MDMAUD_R        0
MDMAUD_F        0
MDMAUD_A        0
GEOCODE2      319
MAILCODE        0
Length: 74, dtype: int64

#### Excluding variables `OSOURCE` and `ZIP`.

In [18]:
drop_list = ['OSOURCE', 'ZIP']
drop_list

['OSOURCE', 'ZIP']

#### Identifying columns that have over 85% missing values and removing them.

In [19]:
def filter_nulls(categorical, perc=0.85):
    nulls_percent_categorical = pd.DataFrame(categorical.isna().sum() / len(categorical)).reset_index()
    nulls_percent_categorical.columns = ['column_name', 'nulls_percentage']
    columns_above_threshold = nulls_percent_categorical[nulls_percent_categorical['nulls_percentage'] > perc]
    drop_columns_list = list(columns_above_threshold['column_name'])
    return drop_columns_list

drop_columns = filter_nulls(categorical, perc=0.85)
drop_columns

['PVASTATE',
 'RECINHSE',
 'RECP3',
 'RECPGVG',
 'RECSWEEP',
 'CHILD03',
 'CHILD07',
 'CHILD12',
 'CHILD18',
 'SOLP3',
 'SOLIH',
 'MAJOR',
 'COLLECT1',
 'VETERANS',
 'BIBLE',
 'CATLG',
 'HOMEE',
 'CDPLAY',
 'STEREO',
 'PCOWNERS',
 'PHOTO',
 'CRAFTS',
 'FISHER',
 'GARDENIN',
 'BOATS',
 'WALKER',
 'KIDSTUFF',
 'CARDS',
 'PLATES']

In [20]:
categorical.drop(columns=drop_columns, axis=1, inplace=True)
categorical.head()

OSOURCE STATE    ZIP NOEXCH MDMAUD DOMAIN CLUSTER AGEFLAG HOMEOWNR GENDER  \
0     GRI    IL  61081      0   XXXX     T2      36     NaN      NaN      F   
1     BOA    CA  91326      0   XXXX     S1      14       E        H      M   
2     AMH    NC  27017      0   XXXX     R2      43     NaN        U      M   
3     BRY    CA  95953      0   XXXX     R2      44       E        U      F   
4     NaN    FL  33176      0   XXXX     S2      16       E        H      F   

  DATASRCE GEOCODE PETS LIFESRC PEPSTRFL RFA_2 RFA_3 RFA_4 RFA_5 RFA_6 RFA_7  \
0      NaN     NaN  NaN     NaN        X   L4E   S4E   S4E   S4E   S4E   S4E   
1        3       2  NaN     NaN      NaN   L2G   A2G   A2G   A2G   A2G   A1E   
2        3     NaN  NaN     NaN        X   L4E   S4E   S4E   S4E   S4E   S4F   
3        3     NaN  NaN     NaN        X   L4E   S4E   S4E   S4E   S4E   S4E   
4        3     NaN  NaN       3      NaN   L2F   A2F   A2F   A2F   A1D   I2D   

  RFA_8 RFA_9 RFA_10 RFA_11 RFA_12 RFA_13 RFA_14 RFA_15 RFA_16 RFA_17 RFA_18  \
0   S4E   S4E    S4E    S4E    S4E    S4E    S4E    S4E    S4E    S4E    S4E   
1   A1E   A1E    A1E    A1E    A1E    NaN    NaN    NaN    L1E    NaN    NaN   
2   S4F   S4F    NaN    S4F    S4F    S4F    S4F    S4F    S4F    NaN    S4D   
3   S4E   S4E    NaN    S4E    S4E    S4E    S4E    S4E    S4E    S2D    S2D   
4   A1E   A1E    L1D    A1E    A1E    L1D    L3D    NaN    L3D    A2D    A2D   

  RFA_19 RFA_20 RFA_21 RFA_22 RFA_23 RFA_24 RFA_2R RFA_2A MDMAUD_R MDMAUD_F  \
0    S4E    S4E    S4E    S4E    S4E    S4E      L      E        X        X   
1    N1E    N1E    N1E    N1E    NaN    F1E      L      G        X        X   
2    S4D    NaN    NaN    S4D    S4D    S3D      L      E        X        X   
3    A1D    A1D    A1D    A1D    NaN    NaN      L      E        X        X   
4    A3D    A3D    A3D    I4E    A3D    A3D      L      F        X        X   

  MDMAUD_A GEOCODE2 MAILCODE  
0        X        C           
1        X        A           
2        X        C           
3        X        C           
4        X        A

#### Reducing the number of categories in the column `GENDER`.

In [21]:
categorical['GENDER'].value_counts()

F    51277
M    39094
U     1715
J      365
C        2
A        2
Name: GENDER, dtype: int64

In [22]:
replace_dict = {'U': 'Other', 'J': 'Other', 'C': 'Other', 'A': 'Other'}
categorical['GENDER'] = df['GENDER'].replace(replace_dict)
categorical['GENDER'].value_counts()

F        51277
M        39094
Other     2084
Name: GENDER, dtype: int64

#### On the **numerical** columns:

In [23]:
df = pd.concat([categorical, numerical, Y], axis=1)
df.shape

(95412, 452)

### Activities

**Use the method value_counts on the columns `MAILCODE`, `NOEXCH`, and `MDMAUD` and check the proportion of category representation in those columns.** Since there is a huge imbalance in the representation of categories, we will add those columns to the drop_list.

In [24]:
df['MAILCODE'].value_counts()

     94013
B     1399
Name: MAILCODE, dtype: int64

In [25]:
df['NOEXCH'].value_counts()

0    61203
0    33882
1      195
1       90
X       35
Name: NOEXCH, dtype: int64

In [26]:
df['NOEXCH'].unique()

array(['0', '1', 'X', 0, 1, nan], dtype=object)

Assuming '0' and 0 as well as '1' and 1 are the same category, there is a high imbalance towards 0.

In [27]:
df['MDMAUD'].value_counts()

XXXX    95118
C1CM       65
L1CM       44
I1CM       37
D2CM       28
C2CM       24
D1CM       20
L2CM       15
L1LM        8
C1LM        8
I2CM        7
D5CM        5
D5MM        5
D2MM        4
C5CM        3
C2MM        3
C2LM        3
D5TM        3
I1LM        3
I5CM        1
C1MM        1
I5MM        1
C5MM        1
I2MM        1
L1MM        1
L2LM        1
C5TM        1
L2TM        1
Name: MDMAUD, dtype: int64

In [28]:
drop_list = ['OSOURCE', 'ZIP']
new_items = ['MAILCODE', 'NOEXCH', 'MDMAUD']
drop_list.extend(new_items)
drop_list

['OSOURCE', 'ZIP', 'MAILCODE', 'NOEXCH', 'MDMAUD']

**Replace null values in the columns `DATASRCE` and `GEOCODE2`.**

In [29]:
df['DATASRCE'].unique()

array([nan, '3', '1', '2'], dtype=object)

`DATASRCE`: Source of Overlay Data: the third-party data source the donor matched against. 1=MetroMail, 2=Polk, 3=Both

In order to not introduce bias, I'll check for correlation of the feature with the targets. If correlation is low, best would be to drop the feature. Unfortunately, earlier I already saw that none of the independent features show a high correlation with either one of the targets. Therefore, I do assign a new category 'Unknown' to the null values. 

In [30]:
df['DATASRCE'] = df['DATASRCE'].replace(np.nan, 'Unknown')
df['DATASRCE'].value_counts()

3          43549
2          23455
Unknown    21280
1           7128
Name: DATASRCE, dtype: int64

**Remove the columns starting with `ADATE_`.** We are assuming that the date when the previous mail was done is not significant in the respondents decision to give donation. They may or may not even remember when they received the mail in the previous years. 

In [31]:
df.drop(columns=df.filter(regex=r'^ADATE_', axis=1).columns, inplace=True)

In [32]:
df.shape

(95412, 429)

### Lab Feature Engineering

#### Checking for null values in the numerical columns.

In [33]:
numerical = df.select_dtypes(np.number)
numerical = numerical.drop(columns = ['TARGET_B', 'TARGET_D'])
numerical.head()

ODATEDW  TCODE   DOB   AGE  NUMCHLD  INCOME  WEALTH1  HIT  MBCRAFT  \
0     8901      0  3712  60.0      NaN     NaN      NaN    0      NaN   
1     9401      1  5202  46.0      1.0     6.0      9.0   16      0.0   
2     9001      1     0   NaN      NaN     3.0      1.0    2      0.0   
3     8701      0  2801  70.0      NaN     1.0      4.0    2      0.0   
4     8601      0  2001  78.0      1.0     3.0      2.0   60      1.0   

   MBGARDEN  MBBOOKS  MBCOLECT  MAGFAML  MAGFEM  MAGMALE  PUBGARDN  PUBCULIN  \
0       NaN      NaN       NaN      NaN     NaN      NaN       NaN       NaN   
1       0.0      3.0       1.0      1.0     1.0      0.0       0.0       0.0   
2       0.0      1.0       0.0      0.0     0.0      0.0       0.0       0.0   
3       0.0      0.0       0.0      0.0     0.0      0.0       0.0       0.0   
4       0.0      9.0       0.0      4.0     1.0      0.0       0.0       0.0   

   PUBHLTH  PUBDOITY  PUBNEWFN  PUBPHOTO  PUBOPP  MALEMILI  MALEVET  VIETVETS  \
0      NaN       NaN       NaN       NaN     NaN         0       39        34   
1      2.0       0.0       3.0       0.0     0.0         0       15        55   
2      1.0       0.0       0.0       0.0     0.0         0       20        29   
3      0.0       0.0       0.0       0.0     1.0         0       23        14   
4      4.0       0.0       1.0       0.0     1.0         1       28         9   

   WWIIVETS  LOCALGOV  STATEGOV  FEDGOV  WEALTH2  POP901  POP902  POP903  \
0        18        10         2       1      5.0     992     264     332   
1        11         6         2       1      9.0    3611     940     998   
2        33         6         8       1      1.0    7001    2040    2669   
3        31         3         0       3      0.0     640     160     219   
4        53        26         3       2      NaN    2520     627     761   

   POP90C1  POP90C2  POP90C3  POP90C4  POP90C5  ETH1  ETH2  ETH3  ETH4  ETH5  \
0        0       35       65       47       53    92     1     0     0    11   
1       99        0        0       50       50    67     0     0    31     6   
2        0        2       98       49       51    96     2     0     0     2   
3        0        8       92       54       46    61     0     0    11    32   
4       99        0        0       46       54     2    98     0     0     1   

   ETH6  ETH7  ETH8  ETH9  ETH10  ETH11  ETH12  ETH13  ETH14  ETH15  ETH16  \
0     0     0     0     0      0      0      0     11      0      0      0   
1     4     2     6     4     14      0      0      2      0      1      4   
2     0     0     0     0      0      0      0      2      0      0      0   
3     6     2     0     0      0      0      0     31      0      0      1   
4     0     0     0     0      0      0      0      0      0      0      0   

   AGE901  AGE902  AGE903  AGE904  AGE905  AGE906  AGE907  CHIL1  CHIL2  \
0      39      48      51      40      50      54      25     31     42   
1      34      41      43      32      42      45      32     33     46   
2      35      43      46      37      45      49      23     35     40   
3      32      40      44      34      43      47      25     45     35   
4      33      45      50      36      46      50      27     34     43   

   CHIL3  AGEC1  AGEC2  AGEC3  AGEC4  AGEC5  AGEC6  AGEC7  CHILC1  CHILC2  \
0     27     11     14     18     17     13     11     15      12      11   
1     21     13     14     33     23     10      4      2      11      16   
2     25     13     20     19     16     13     10      8      15      14   
3     20     15     25     17     17     12      7      7      20      17   
4     23     14     21     13     15     20     12      5      13      15   

   CHILC3  CHILC4  CHILC5  HHAGE1  HHAGE2  HHAGE3  HHN1  HHN2  HHN3  HHN4  \
0      34      25      18      26      10      23    18    33    49    28   
1      36      22      15      12       1       5     4    21    75    55   
2      30      22      19      25      10   

In [34]:
from IPython.display import display

nan_counts = numerical.isna().sum()
nan_counts = nan_counts[nan_counts > 0]
nan_df = pd.DataFrame({'Column': nan_counts.index, 'NaN Count': nan_counts.values})
with pd.option_context('display.max_rows', None):
    display(nan_df)

Column  NaN Count
0        AGE      23665
1    NUMCHLD      83026
2     INCOME      21286
3    WEALTH1      44732
4    MBCRAFT      52854
5   MBGARDEN      52854
6    MBBOOKS      52854
7   MBCOLECT      52914
8    MAGFAML      52854
9     MAGFEM      52854
10   MAGMALE      52854
11  PUBGARDN      52854
12  PUBCULIN      52854
13   PUBHLTH      52854
14  PUBDOITY      52854
15  PUBNEWFN      52854
16  PUBPHOTO      52854
17    PUBOPP      52854
18   WEALTH2      43823
19       MSA        132
20       ADI        132
21       DMA        132
22   RDATE_3      95170
23   RDATE_4      95131
24   RDATE_5      95403
25   RDATE_6      94636
26   RDATE_7      86517
27   RDATE_8      73940
28   RDATE_9      78678
29  RDATE_10      84951
30  RDATE_11      80672
31  RDATE_12      69712
32  RDATE_13      83162
33  RDATE_14      72095
34  RDATE_15      88150
35  RDATE_16      68418
36  RDATE_17      86011
37  RDATE_18      75634
38  RDATE_19      79535
39  RDATE_20      87524
40  RDATE_21      85899
41  RDATE_22      74539
42  RDATE_23      87553
43  RDATE_24      77674
44   RAMNT_3      95170
45   RAMNT_4      95131
46   RAMNT_5      95403
47   RAMNT_6      94636
48   RAMNT_7      86517
49   RAMNT_8      73940
50   RAMNT_9      78678
51  RAMNT_10      84951
52  RAMNT_11      80672
53  RAMNT_12      69712
54  RAMNT_13      83162
55  RAMNT_14      72095
56  RAMNT_15      88150
57  RAMNT_16      68418
58  RAMNT_17      86011
59  RAMNT_18      75634
60  RAMNT_19      79535
61  RAMNT_20      87524
62  RAMNT_21      85899
63  RAMNT_22      74539
64  RAMNT_23      87553
65  RAMNT_24      77674
66  NEXTDATE       9973
67   TIMELAG       9973
68  CLUSTER2        132

There is still many numerical columns with lots of NaNs, the removal in the categorical ones didn't seem to have a big effect here. I will run again the filter_nulls function with a larger treshold on the numerical data to rmeove more of them than before. 

In [35]:
len(df[df['TARGET_D'] >= 50])

114

In [36]:
drop_columns = filter_nulls(numerical, perc=0.25)
drop_columns

['NUMCHLD',
 'WEALTH1',
 'MBCRAFT',
 'MBGARDEN',
 'MBBOOKS',
 'MBCOLECT',
 'MAGFAML',
 'MAGFEM',
 'MAGMALE',
 'PUBGARDN',
 'PUBCULIN',
 'PUBHLTH',
 'PUBDOITY',
 'PUBNEWFN',
 'PUBPHOTO',
 'PUBOPP',
 'WEALTH2',
 'RDATE_3',
 'RDATE_4',
 'RDATE_5',
 'RDATE_6',
 'RDATE_7',
 'RDATE_8',
 'RDATE_9',
 'RDATE_10',
 'RDATE_11',
 'RDATE_12',
 'RDATE_13',
 'RDATE_14',
 'RDATE_15',
 'RDATE_16',
 'RDATE_17',
 'RDATE_18',
 'RDATE_19',
 'RDATE_20',
 'RDATE_21',
 'RDATE_22',
 'RDATE_23',
 'RDATE_24',
 'RAMNT_3',
 'RAMNT_4',
 'RAMNT_5',
 'RAMNT_6',
 'RAMNT_7',
 'RAMNT_8',
 'RAMNT_9',
 'RAMNT_10',
 'RAMNT_11',
 'RAMNT_12',
 'RAMNT_13',
 'RAMNT_14',
 'RAMNT_15',
 'RAMNT_16',
 'RAMNT_17',
 'RAMNT_18',
 'RAMNT_19',
 'RAMNT_20',
 'RAMNT_21',
 'RAMNT_22',
 'RAMNT_23',
 'RAMNT_24']

Wealth might play a role when donating, therefore I'll keep the features related.
`WEALTH`: wealth rating of donor ususally including real estate, financial, and other wealth indicators
`WEALTH2`: relative wealth within each state

In [37]:
cols = ['WEALTH1', 'WEALTH2']
for item in cols:
    drop_columns.remove(item) 

In [38]:
df_less_nulls = df.drop(columns=drop_columns)

In [39]:
# To make sure that no rows of high-amount-donors were dropped, I compare the amount of rows with the amount before dropping.
len(df_less_nulls[df_less_nulls['TARGET_D'] >= 50])

114

In [40]:
df = df_less_nulls
df.shape

(95412, 370)

**Cleaning the columns `GEOCODE2`, `WEALTH1`, `ADI`, `DMA`, and `MSA`.**

I go back to working on the entire dataset with both numerical and categorical data and targets.

In [41]:
df['GEOCODE2'].unique()

array(['C', 'A', 'D', 'B', nan], dtype=object)

In [42]:
df['GEOCODE2'].dropna()

0        C
1        A
2        C
3        C
4        A
        ..
95407    C
95408    A
95409    B
95410    A
95411    C
Name: GEOCODE2, Length: 95093, dtype: object

In [43]:
df['WEALTH1'].isna().sum()

44732

In [44]:
df['WEALTH1'].value_counts()

9.0    7585
8.0    6793
7.0    6198
6.0    5825
5.0    5280
4.0    4810
3.0    4237
2.0    4085
1.0    3454
0.0    2413
Name: WEALTH1, dtype: int64

We are looking at wealth rating across (a sample of) society, hence I am assuming that the proportions of the existing categories are representative of the distribution of the missing values. I assign the null values proportionally to each existing category.

In [45]:
# Calculate the proportion of each existing category.
proportions = df['WEALTH1'].value_counts(normalize=True)
proportions

9.0    0.149665
8.0    0.134037
7.0    0.122297
6.0    0.114937
5.0    0.104183
4.0    0.094909
3.0    0.083603
2.0    0.080604
1.0    0.068153
0.0    0.047612
Name: WEALTH1, dtype: float64

In [46]:
counts = df['WEALTH1'].value_counts(normalize=True)
df['WEALTH1'].fillna(pd.Series(np.random.choice(counts.index, size=len(df.index), p=counts)), inplace=True)
df['WEALTH1'].value_counts()

9.0    14245
8.0    12842
7.0    11548
6.0    11021
5.0     9889
4.0     9109
3.0     8031
2.0     7715
1.0     6423
0.0     4589
Name: WEALTH1, dtype: int64

`ADI`, `DMA`, `MSA` are codes for geographic identification in the US
- ADI code: Area Deprivation Index code is a measure of socioeconomic deprivation at the neighborhood level. It is used to identify areas with high levels of poverty, unemployment, and other indicators of disadvantage.
- DMA code: Designated Market Area code is a geographic area defined by Nielsen Media Research to identify television viewing markets. DMAs are used by advertisers and broadcasters to determine the reach of television advertising.
- MSA code: Metropolitan Statistical Area code is a designation to refer to a delineation consisting of a city and its suburbs. MSAs are used to group counties and cities into specific geographic areas for population censuses and compilations of related statistical data.

In [47]:
df['ADI'].isna().sum()

132

In [48]:
df['DMA'].isna().sum()

132

In [49]:
nan_rows = df[df['MSA'].isna()]
nan_rows

OSOURCE STATE     ZIP NOEXCH MDMAUD DOMAIN CLUSTER AGEFLAG HOMEOWNR  \
577       BHG    FL   33756      0   XXXX     U3       9       E        H   
1119      SPN    FL   34642      0   XXXX     C2      29     NaN      NaN   
2250      DNA    GA   31535      0   XXXX     T2      36       E        U   
3326      SYN    GA   31217      0   XXXX     C3      32     NaN      NaN   
5558      SSS    NC   28625      0   XXXX     T2      36       E        H   
...       ...   ...     ...    ...    ...    ...     ...     ...      ...   
90627     L01    WA   98375      0   XXXX     C2      27       E        H   
90993     LIF    SC  59887-      0   XXXX     C2      26       E        U   
92870     BHG    FL   34624      0   XXXX     U3       8       E        H   
93624     HAN    NC   28370      0   XXXX     C1      23       E        U   
94604     APP    AZ   86446      0   XXXX     R2      45     NaN        H   

      GENDER DATASRCE GEOCODE PETS LIFESRC PEPSTRFL RFA_2 RFA_3 RFA_4 RFA_5  \
577        F        2      14  NaN       3        X   L1E   A1E   A1E   NaN   
1119       F  Unknown      14  NaN     NaN        X   L1F   A1F   A1F   NaN   
2250       M        3     NaN    Y       3      NaN   L1G   A1G   A1G   A1G   
3326       M  Unknown     NaN  NaN       1        X   L3G   N2G   N2G   N2G   
5558       M        3     NaN  NaN       3        X   L4E   S4E   S4E   S4E   
...      ...      ...     ...  ...     ...      ...   ...   ...   ...   ...   
90627      M        2     NaN  NaN     NaN      NaN   L1F   A1F   A1F   A1F   
90993      F        3     NaN  NaN     NaN      NaN   L1F   A1F   A1F   NaN   
92870      M        3     NaN  NaN       2        X   L4E   S4E   S4E   S4E   
93624      M        2      14  NaN     NaN      NaN   L1F   A1F   A1F   NaN   
94604      M        2     NaN  NaN     NaN      NaN   L2G   A2G   A2G   A2G   

      RFA_6 RFA_7 RFA_8 RFA_9 RFA_10 RFA_11 RFA_12 RFA_13 RFA_14 RFA_15  \
577     S2E   S2E   S2E   S3E    S3E    S3E    S3E    S4E    S4D    S4D   
1119    A1F   L1E   L1E   L2E    NaN    L2E    L2E    NaN    L2E    NaN   
2250    NaN   NaN   NaN   NaN    NaN    I1G    I1G    NaN    NaN    NaN   
3326    N2G   N2G   N2G   F1F    NaN    F1F    F1F    NaN    P1F    NaN   
5558    S4E   A4D   A4D   A4D    A4D    A4D    A4D    A4D    A4D    A4D   
...     ...   ...   ...   ...    ...    ...    ...    ...    ...    ...   
90627   A1F   A1F   A1F   A2F    A2F    A2F    A2F    A1F    A1F    NaN   
90993   N2F   N1E   N1E   N1E    N1E    N1E    N1E    F1E    F1E    NaN   
92870   S4E   S4E   S4E   S4E    S4E    S4E    S4E    S4D    S4D    S4D   
93624   A1F   A1F   A1F   A1F    A1F    A1F    A1F    A1F    A1F    A2F   
94604   A2G   A1G   A1G   L1G    NaN    L1G    L1G    NaN    L1G    NaN   

      RFA_16 RFA_17 RFA_18 RFA_19 RFA_20 RFA_21 RFA_22 RFA_23 RFA_24 RFA_2R  \
577      S3D    S4D    S4D    A4D    NaN    NaN    A4D    A3D    A3D      L   
1119     L3E    A1E    A1E    S2E    S2E    S2E    S2E    S2E    S2E      L   
2250     I2G    L1G    L1G    L1G    NaN    L1G    L1G    NaN    L2G      L   
3326     NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN      L   
5558     A3D    NaN    A1D    A1D    A1D    A1D    A1D    NaN    A1D      L   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
90627    N2F    N2F    N2F    N1E    NaN    NaN    F1E    NaN    F1E      L   
90993    NaN    P1E    P1E    NaN    NaN    NaN    NaN    NaN    NaN      L   
92870    S4D    S3D    S3D    S3C    NaN    NaN    S2C    NaN    A2C      L   
93624    A2F    A1E    A1E    A1E    A1E    A1E    A1E    A1E    NaN      L   
94604    L1G    A1G    A1G    A1G    A1G    A1G    A1G    A1G    A1G      L   

      RFA_2A MDMAUD_R MDMAUD_F MDMAUD_A GEOCODE2 MAILCODE  ODATEDW  TCODE  \
577        E        X        X        X      NaN              8601      2   
1119       F        X        X        X      NaN              8601      0   
2250       G        X        X        X      NaN   

All three features have NaNs in the exact same rows. And there are only 132 in total. Therefore I will drop all NaNs.

In [50]:
df.dropna(subset=['ADI'], inplace=True)

In [51]:
# To confirm.
nan_rows = df[df['DMA'].isna()]
nan_rows

Empty DataFrame
Columns: [OSOURCE, STATE, ZIP, NOEXCH, MDMAUD, DOMAIN, CLUSTER, AGEFLAG, HOMEOWNR, GENDER, DATASRCE, GEOCODE, PETS, LIFESRC, PEPSTRFL, RFA_2, RFA_3, RFA_4, RFA_5, RFA_6, RFA_7, RFA_8, RFA_9, RFA_10, RFA_11, RFA_12, RFA_13, RFA_14, RFA_15, RFA_16, RFA_17, RFA_18, RFA_19, RFA_20, RFA_21, RFA_22, RFA_23, RFA_24, RFA_2R, RFA_2A, MDMAUD_R, MDMAUD_F, MDMAUD_A, GEOCODE2, MAILCODE, ODATEDW, TCODE, DOB, AGE, INCOME, WEALTH1, HIT, MALEMILI, MALEVET, VIETVETS, WWIIVETS, LOCALGOV, STATEGOV, FEDGOV, WEALTH2, POP901, POP902, POP903, POP90C1, POP90C2, POP90C3, POP90C4, POP90C5, ETH1, ETH2, ETH3, ETH4, ETH5, ETH6, ETH7, ETH8, ETH9, ETH10, ETH11, ETH12, ETH13, ETH14, ETH15, ETH16, AGE901, AGE902, AGE903, AGE904, AGE905, AGE906, AGE907, CHIL1, CHIL2, CHIL3, AGEC1, AGEC2, AGEC3, AGEC4, AGEC5, AGEC6, ...]
Index: []

Before moving to EDA I drop the features from the drop_list from earlier.

In [52]:
drop_list

['OSOURCE', 'ZIP', 'MAILCODE', 'NOEXCH', 'MDMAUD']

In [53]:
df.drop(columns=drop_list, inplace=True)

In [54]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 95280 entries, 0 to 95411
Columns: 365 entries, STATE to TARGET_D
dtypes: float64(16), int64(309), object(40)
memory usage: 266.1+ MB


In [55]:
df.set_index('CONTROLN')

STATE DOMAIN CLUSTER AGEFLAG HOMEOWNR GENDER DATASRCE GEOCODE PETS  \
CONTROLN                                                                      
95515       IL     T2      36     NaN      NaN      F  Unknown     NaN  NaN   
148535      CA     S1      14       E        H      M        3       2  NaN   
15078       NC     R2      43     NaN        U      M        3     NaN  NaN   
172556      CA     R2      44       E        U      F        3     NaN  NaN   
7112        FL     S2      16       E        H      F        3     NaN  NaN   
...        ...    ...     ...     ...      ...    ...      ...     ...  ...   
184568      AK     C2      27     NaN      NaN      M  Unknown     NaN  NaN   
122706      TX     C1      24       E        H      M        3     NaN  NaN   
189641      MI     C3      30     NaN      NaN      M  Unknown     NaN  NaN   
4693        CA     C1      24       E        H      F        2       4  NaN   
185114      NC     C1      24       E        U      F        3       3  NaN   

         LIFESRC PEPSTRFL RFA_2 RFA_3 RFA_4 RFA_5 RFA_6 RFA_7 RFA_8 RFA_9  \
CONTROLN                                                                    
95515        NaN        X   L4E   S4E   S4E   S4E   S4E   S4E   S4E   S4E   
148535       NaN      NaN   L2G   A2G   A2G   A2G   A2G   A1E   A1E   A1E   
15078        NaN        X   L4E   S4E   S4E   S4E   S4E   S4F   S4F   S4F   
172556       NaN        X   L4E   S4E   S4E   S4E   S4E   S4E   S4E   S4E   
7112           3      NaN   L2F   A2F   A2F   A2F   A1D   I2D   A1E   A1E   
...          ...      ...   ...   ...   ...   ...   ...   ...   ...   ...   
184568       NaN      NaN   L1G   F1G   F1G   NaN   F1G   P1G   P1G   NaN   
122706       NaN      NaN   L1F   NaN   NaN   NaN   NaN   P1F   P1F   NaN   
189641       NaN        X   L3E   S4E   S4E   S3E   S3E   NaN   A2E   N3E   
4693           1        X   L4F   S4F   A3F   S4F   S4F   S4F   S4F   S4F   
185114       NaN        X   L1G   A1G   A1G   A1G   A1G   L3G   L3G   L4G   

         RFA_10 RFA_11 RFA_12 RFA_13 RFA_14 RFA_15 RFA_16 RFA_17 RFA_18  \
CONTROLN                                                                  
95515       S4E    S4E    S4E    S4E    S4E    S4E    S4E    S4E    S4E   
148535      A1E    A1E    A1E    NaN    NaN    NaN    L1E    NaN    NaN   
15078       NaN    S4F    S4F    S4F    S4F    S4F    S4F    NaN    S4D   
172556      NaN    S4E    S4E    S4E    S4E    S4E    S4E    S2D    S2D   
7112        L1D    A1E    A1E    L1D    L3D    NaN    L3D    A2D    A2D   
...         ...    ...    ...    ...    ...    ...    ...    ...    ...   
184568      NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
122706      NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
189641      N3E    N3E    N3E    N3E    N3E    N2E    F1D    NaN    F1D   
4693        S4F    S4F    S4F    S4F    S4F    S4F    S3F    S2F    S2F   
185114      A1G    A1G    A1G    A1G    A1G    A1G    S2G    S3G    S3G   

         RFA_19 RFA_20 RFA_21 RFA_22 RFA_23 RFA_24 RFA_2R RFA_2A MDMAUD_R  \
CONTROLN                                                                    
95515       S4E    S4E    S4E    S4E    S4E    S4E      L      E        X   
148535      N1E    N1E    N1E    N1E    NaN    F1E      L      G        X   
15078       S4D    NaN    NaN    S4D    S4D    S3D      L      E        X   
172556      A1D    A1D    A1D    A1D    NaN    NaN      L      E        X   
7112        A3D    A3D    A3D    I4E    A3D    A3D      L      F        X   
...         ...    ...    ...    ...    ...    ...    ...    ...      ...   
184568      NaN    NaN    NaN    NaN    NaN    NaN      L      G        X   
122706      NaN    NaN    NaN    NaN    NaN    NaN      L      F        X   
189641      NaN    NaN    P1D    P1D    NaN    NaN      L      E        X   
4693        A1F    A1F    A1F    A1F    S2F    S3F      L      F        X   
185114      S4G    S3G    S3G    S3G    S3G    S3G      L      G        C   

 

The dataset has still too many independent features and too many null values. To quickly reduce the amount of features I will drop those columns with a very high number of categories. Neither no variance nor extremely high variance is helpful in categorical columns.

In [56]:
remove_cols = []

for col in df:
    if len(df[col].unique()) > 50:
        remove_cols.append(col)

num_remove_cols = len(remove_cols)
print("Number of columns to be removed:", num_remove_cols)

Number of columns to be removed: 281


In [59]:
df = df.drop(columns=remove_cols)

In [60]:
df.head()

DOMAIN AGEFLAG HOMEOWNR GENDER DATASRCE GEOCODE PETS LIFESRC PEPSTRFL RFA_2  \
0     T2     NaN      NaN      F  Unknown     NaN  NaN     NaN        X   L4E   
1     S1       E        H      M        3       2  NaN     NaN      NaN   L2G   
2     R2     NaN        U      M        3     NaN  NaN     NaN        X   L4E   
3     R2       E        U      F        3     NaN  NaN     NaN        X   L4E   
4     S2       E        H      F        3     NaN  NaN       3      NaN   L2F   

  RFA_5 RFA_15 RFA_2R RFA_2A MDMAUD_R MDMAUD_F MDMAUD_A GEOCODE2  INCOME  \
0   S4E    S4E      L      E        X        X        X        C     NaN   
1   A2G    NaN      L      G        X        X        X        A     6.0   
2   S4E    S4F      L      E        X        X        X        C     3.0   
3   S4E    S4E      L      E        X        X        X        C     1.0   
4   A2F    NaN      L      F        X        X        X        A     3.0   

   WEALTH1  WEALTH2  ETH6  ETH10  ETH11  ETH12  ETH14  AGEC4  AGEC5  CHILC2  \
0      2.0      5.0     0      0      0      0      0     17     13      11   
1      9.0      9.0     4     14      0      0      0     23     10      16   
2      1.0      1.0     0      0      0      0      0     16     13      14   
3      4.0      0.0     6      0      0      0      0     17     12      17   
4      2.0      NaN     0      0      0      0      0     15     20      15   

   HV3  HV4  HHD8  ETHC4  RHP3  RHP4  HUPA7  IC11  IC12  IC13  IC21  IC22  \
0    3    2     2      0    15     4      0     1     0     0     0     0   
1   12   10     1      0    21     5      0    14    13    10    14    10   
2    2    1     2      0    14     5      6     2     0     0     0     0   
3    2    1     3      0    15     5      5     5     0     0     0     0   
4    4    3     3     21    16     6      0     4     0     0     0     0   

   TPE5  TPE6  TPE7  OCC3  OCC6  OCC7  OCC12  OCC13  EIC5  EIC6  EIC7  EIC9  \
0     0     0     0     8     0     3      5      4     1     0     7     6   
1     0     0     0     4     2     1      0      3     1     2     8    11   
2     0     0     0     2     0     1      5      7     4     3     2     3   
3     0     0     0     0     0     0      0      0     0     0     0     3   
4     0     0     0     2     3     7      7      2     5     2     2     2   

   EIC10  EIC11  EIC12  OEDC7  EC6  SEC3  AFC3  AFC6  ANC3  ANC5  ANC6  ANC7  \
0      5      0      4      0    8     3     0     0     0     0     0     5   
1      4      3      4      2    7     4     0     0     0     1     1     2   
2      3      2      0      1    6     1     0     0     1     0     0     7   
3      3      6      0      0   10     1     0     0     0     0     0     4   
4      7      6      4      1    3     1     1     4     0     0     0     0   

   ANC8  ANC11  ANC12  ANC13  ANC14  ANC15  HC1  HC15  MHUC1  MHUC2  AC1  AC2  \
0     1      0      0      0      2      0   13     0      6      2   10    7   
1     3      0      3      0      0      0    5     0     20      4    6    5   
2     0      0      0      1      0      0    9     0      9      2    6    5   
3     0      3      0      0      0      0    4     0     16      2    6    6   
4     0      0      0      0      0      0   25     0      6      2    7   11   

   MAXADATE  CARDPM12  CARDGIFT  LASTDATE  HPHONE_D  RFA_2F  TARGET_B  
0      9702         6        14      9512         0       4         0  
1      9702         6         1      9512         0       2         0  
2      9702         6        14      9512         1       4         0  
3      9702         6         7      9512         1       4         0  
4      9702        10         8      9601         1       2         0

There is still a lot of null values, where we would have to explore specific features further to deal with before concluding the EDA.